linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/InternalModuleIDGenerator.scala

```diff
@@ -62,22 +88,79 @@ private[modulesplitter] final class InternalModuleIDGenerator(
      *
      * Note that this is stable, because it does not depend on the order we
      * iterate over nodes.
+     *
+     * To deal with case-insensitive issues, basically we prefix every
+     * uppercase character with a '-', and we prefix every '-' with a '-' to
+     * avoid clashes. However, that is not good enough, since several
+     * uppercase (and titlecase) code points can case-fold to the same
+     * lowercase letter. Therefore, the complete scheme is:
+     *
+     * - ASCII uppercase letters are prefixed with '-'.
+     * - '-' is prefixed with '-'.
+     * - Non-ASCII characters are all prefixed by '-u' followed by the 6
+     *   hexdigits of their codepoint.
+     *
+     * The last rule is far from being optimal, but it is safe. Encountering
+     * non-ASCII characters in class names should be rare anyway.
      */
-    var moduleID = ModuleID(name.nameString)
-    while (avoidSet.contains(moduleID))
-      moduleID = ModuleID(moduleID.id + ".")
-    moduleID
+
+    val builder = new java.lang.StringBuilder
+
+    // First, encode uppercase characters to avoid accidental case-insensitive clashes
+    val originalNameString = name.nameString
+    val originalNameStringLen = originalNameString.length()
+    var i = 0
+    while (i != originalNameStringLen) {
+      val cp = originalNameString.codePointAt(i)
+      if (cp < 0x80) {
+        // ASCII
+        if (cp == '-' || (cp >= 'A' && cp <= 'Z'))
+          builder.append('-')
+        builder.append(cp.toChar)
+        i += 1
+      } else {
+        // Non-ASCII
+        new java.util.Formatter(builder).format("-u%06x", Integer.valueOf(cp))
```

Comment:
Why do we need `Integer.valueOf`? Shouldn't the compiler do that for us?

Reply:
It's a quirk of using Java `Object...` varargs parameters. Normally `Object` in a Java signature is interpreted as `Any`, but not if it is an `Object[]` (which an `Object...` is underneath). In that case it's still `AnyRef`, and an `Int` does not conform to `AnyRef`.